### ISSUE TITLE
Add mysql queries

### ISSUE THREAD

**Issue by brianvoe** [brianvoe/gofakeit#195]
Need to add the ability to generate mysql queries for mass imports

**Comment by petergeorgas** [comment]
What would this entail? I would love to take a crack at this as an opportunity to do more with SQL & Go.

**Comment by brianvoe** [comment]
So we seed our local databases for testing and to have data to be able to pass to the front end so front end developers have data they can use to visualize what the website looks like with actual data in it.

**Comment by petergeorgas** [comment]
So, for example, we would have a function named, say, `InsertQuery` that takes a struct that is to be populated with our fake data, as well as the table name,  and then generates an insertion query to do so? Say the struct was this:

```go
type PersonEntry struct {
	Fname string `fake:"{firstname}"`
	Lname string `fake:"{lastname}"`
	Age int `fake:"{number}"`
}

var p PersonEntry
gofakeit.Struct(&p)
```
and say the function call was `faker.InsertQuery(&p, "PEOPLE")`?

Is this what you're looking for?

**Comment by brianvoe** [comment]
So I would probably think about it in terms of batch inserts.

```mysql
INSERT INTO yourtable VALUES (1,2), (5,5), ...;
```

So just like the json function they would pass in a MysqlOptions and one of the fields would be an array of fields.

https://github.com/brianvoe/gofakeit/blob/master/json.go#L58

**Comment by petergeorgas** [comment]
I get it! That makes sense. I will give this a shot as soon as I have some free time this week.

**Comment by petergeorgas** [comment]
This is what I've got so far: 

```go
type SQLOptions struct {
	Table  string  `json:"table" xml:"table"`
	Fields []Field `json:"fields" xml:"fields"`
}

func SQLInsert(so *SQLOptions) ([]byte, error) {
	return sqlInsertFunc(globalFaker.Rand, so)
}

func sqlInsertFunc(r *rand.Rand, so *SQLOptions) ([]byte, error) {
	if so.Table == "" {
		return nil, errors.New("must provide table name to generate SQL")
	}
	if so.Fields == nil || len(so.Fields) <= 0 {
		return nil, errors.New(("must pass fields in order to generate SQL queries"))
	}

	var sb strings.Builder
	sb.WriteString("INSERT INTO " + so.Table + " VALUES ")

	// Now, we need to add all of our fields
	for _, field := range so.Fields {
		if field.Function == "autoincrement" { // One
			// TODO: We need to do something here still...
			continue
		}

		// Get the function info for the field
		funcInfo := GetFuncLookup(field.Function)
		if funcInfo == nil {
			return nil, errors.New("invalid function, " + field.Function + " does not exist")
		}

		val, err := funcInfo.Generate(r, &field.Params, funcInfo)
		if err != nil {
			return nil, err
		}

		

	}

	return []byte(sb.String()), nil
}
```

What I am not understanding here is how we would do batch inserts when given an array of fields, to me this is looking like this would just end up doing one tuple, no? Would I need an array of arrays of fields?

**Comment by brianvoe** [comment]
Haha these are all great questions! Its also probably why I havent tackled this one yet. That being said as long as you can generate the value you can use the funcInfo to know what the output type is and have a conversion function that looks at the type and converts it to the type that sql supports. Thats where I would start and see how far you get and hit me up if you need anymore help.

**Comment by petergeorgas** [comment]
Is the `ContentType` property of `Info` what indicates the output type? Or is it `Output`?

**Comment by petergeorgas** [comment]
I'm thinking, I would say 95% of the use cases are putting either `TEXT`, `FLOAT, INTEGER, DECIMAL, ETC.`, and `DATE` types into the DB. 

Would something as simple as this suffice for now? I'm missing the DATE stuff right now, but I can probably put something together. Basically, we just add `'` around the string types and convert the numerical types to their values and it should be good. At the end of the day, we're putting this all in one string to return so their types don't matter, most of SQL from what I'm recalling, besides strings and DATEs has nothing surrounding it.
```go
func ConvertType(t string, val interface{}) string {
	switch t {
	case "string":
		return `'` + fmt.Sprintf("%v", t) + `'`
	default: 
		return fmt.Sprintf("%v", t)
	}
}
```

**Comment by petergeorgas** [comment]
Here's what I have so far, what do you think?

[sql.go](https://github.com/petergeorgas/gofakeit/blob/feature/add_sql_queries/sql.go)
[sql_test.go](https://github.com/petergeorgas/gofakeit/blob/feature/add_sql_queries/sql_test.go

**Comment by brianvoe** [comment]
I think those look great. One last thing you may want to do is throw more variations at it to make sure your covering all your bases. Things like floats. Also json can be a column type in mysql so you might want to try that as well.

If those look good and tests run ill get it merged in.

Nice job!

**Comment by petergeorgas** [comment]
Floats/Decimal values are handled by my `ConvertType` function! I will write a few tests to make sure. 

Looking at the [MySQL documentation ](https://dev.mysql.com/doc/refman/8.0/en/json.html), JSON is inserted into a table by just _stringifying_ it, so I can just add the logic to do the same thing we do for the `string` data type as we do for `json` -- wrap the string in `'` characters. I will also add a test for this to make sure.

**Comment by brianvoe** [comment]
Thats a good question. I think that should be the case but youll have to test it out.

**Comment by petergeorgas** [comment]
`Float` works great! I am having a bit of trouble understanding how I would specify I want to generate fake JSON data within a field for my `SQLOptions`.

```go
res, _ := SQLInsert(&SQLOptions{
		Table:      "People",
		EntryCount: 3,
		Fields: []Field{
			{Name: "first_name", Function: "firstname"},
			{Name: "last_name", Function: "lastname"},
			{Name: "age", Function: "number", Params: MapParams{"min": {"1"}, "max": {"99"}}},
		},
	})
```

I know that my `Function` property will be `"json"`, but what do I use for `MapParams`?

**Comment by brianvoe** [comment]
Ya it can get a little tricky putting together a sub json field. The cleanest way without having large amounts of param strings is to just create a simple lookup function with some fields in it and then in your sql test field just add the custom lookup function you added.

Here is an example of adding a custom function. https://github.com/brianvoe/gofakeit/blob/master/lookup_test.go#L11

Let me know if that makes sense to you.

We are really close. If you want to start a pr submission I can start looking at it more closely after you add the json test.

**Comment by petergeorgas** [comment]
I am having some trouble here, and I really can't seem to figure out what I'm doing incorrect. I did something nearly similar to your example, but I'm just getting random jumbled nonsense. This is my lookup function that I created, I'm working on integrating the two together so bear with me please:

```go
func TestMultiSQLInsertJSON(t *testing.T) {

	Seed(12)

	AddFuncLookup("jsonperson", Info{
		Category:    "custom",
		Description: "random JSON of a person",
		Example:     `{"first_name":"Bob", "last_name":"Jones"}`,
		Output:      "[]byte",
		Generate: func(r *rand.Rand, m *MapParams, info *Info) (interface{}, error) {

			v, _ := JSON(&JSONOptions{
				Type:     "object",
				RowCount: 1,
				Fields: []Field{
					{Name: "first_name", Function: "firstname"},
					{Name: "last_name", Function: "lastname"},
				},
			})

			fmt.Printf("Z: %s\n", string(v))

			return v, nil
		},
	})

	type RandomPerson struct {
		FakePerson []byte `fake:{jsonperson}"`
	}

	var f RandomPerson
	Struct(&f)
	fmt.Println(string(f.FakePerson))
	//fmt.Printf("%s\n", string(f.FakePerson))
}
```

```go
v, _ := JSON(&JSONOptions{
				Type:     "object",
				RowCount: 1,
				Fields: []Field{
					{Name: "first_name", Function: "firstname"},
					{Name: "last_name", Function: "lastname"},
				},
			})

**Comment by brianvoe** [comment]
Can you submit a pr and ill pull it down and take a look at it?

**Comment by petergeorgas** [comment]
#201 is the PR! Let me know what I can do!

**Comment by brianvoe** [comment]
Just merged your commit into v6.16.0. I made some updates, cleanups and added the json usage and lookup usage. Once again nice job on this integration!
